flink-runtime/src/main/java/org/apache/flink/runtime/state/IncrementalRemoteKeyedStateHandle.java

```diff
@@ -204,6 +203,21 @@ public long getStateSize() {
         return size;
     }
 
+    @Override
+    public long getIncrementalStateSize() {
+        long size = StateUtil.getStateSize(metaStateHandle);
+
+        for (StreamStateHandle sharedStateHandle : sharedState.values()) {
+            size += sharedStateHandle.getIncrementalStateSize();
```

Comment:
I guess this only works because `PlaceholderStreamStateHandle.getIncrementalStateSize` returns `0`.
But backend isn't requried to return placeholder; in fact, it currently doesn't - without FLINK-25395/ #18297 (In the future, the latter PR quite likely will be reverted I think).

WDYT about computing incremental state size externally (in `RocksIncrementalSnapshotStrategy`) and storing it in metadata?

Reply:
What do you mean storing it in meatadata? Current checkpoint stats shown in UI depend on summarizing state size from state handle. For other state backend, they can return the real incremental state size in `StateHandle#getIncrementalStateSize` method.

Reply:
I mean `IncrementalRemoteKeyedStateHandle` could have a final variable `incrementalSize` and return it from `getIncrementalStateSize`.
To create a handle:
- On TM, it would be computed by `RocksIncrementalSnapshotStrategy`.
- On JM, it would be read from checkpoint metadata by `MetadataV2V3SerializerBase#serializeKeyedStateHandle` (and written).

It shouldn't change after the creation of `IncrementalRemoteKeyedStateHandle`, right?

Reply:
Already persist in the new IncrementalRemoteStateHandle